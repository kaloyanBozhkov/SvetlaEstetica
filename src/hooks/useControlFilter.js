import { useState } from 'react'

// controller for Filter component
const useControlFilter = ({ categories = [], maxPrice }) => {
    const [selectedCategory, setSelectedCategory] = useState(categories[0]?.id || -1)
    const [currentPrice, setPriceRange] = useState(maxPrice)
    return {
        selectedCategory,
        currentPrice,
        setPriceRange,
        setSelectedCategory
    }
}

export default useControlFilter