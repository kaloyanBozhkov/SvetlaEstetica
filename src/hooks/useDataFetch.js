import { useEffect, useState } from 'react'

/**
 * @param {fetch func} fetcher => fetch
 * @param {func} checker => predicate acting as gate for fetch
 * @returns {[data, loading, error]}
 */
const useDataFetch = (fetcher, checker) => {
    const [loading, setLoading] = useState()
    const [error, setError] = useState()
    const [data, setData] = useState()

    useEffect(() => {

        if (checker()) {

            // set loading
            setLoading(true)

            // start async promise, and update data or error when fulfilled (handle eiuther rejected or resolved)
            fetcher()
                .then(resp => resp.json())
                .then(setData)
                .catch(setError)
        }
    }, [fetcher, checker])


    return [data, loading, error]
}

export default useDataFetch