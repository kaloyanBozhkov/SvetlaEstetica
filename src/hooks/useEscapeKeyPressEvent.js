import { useEffect } from 'react'

function useEscapeKeyPressEvent(fn) {
    useEffect(() => {
        const handleEvent = (event) => {
            if (event.key === 'Escape') fn()
        }
        window.addEventListener('keydown', handleEvent)
        return () => window.removeEventListener('keydown', handleEvent)
    }, [fn])
}

export default useEscapeKeyPressEvent
