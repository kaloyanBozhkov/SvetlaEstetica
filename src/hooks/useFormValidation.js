import { useState } from 'react'

const cloneDefaultValues = (defaultValues) => {

    // this will overlook functions, regexp etc.. so we gotta apply the custom rules again
    const deepClone = JSON.parse(JSON.stringify(defaultValues))

    // get { fieldName: customRules } from defaultValues obj
    Object.keys(defaultValues).forEach((formName) => {
        const customRules = defaultValues[formName]?.validation?.customRules || null
        if (defaultValues[formName]?.validation) {
            deepClone[formName].validation.customRules = customRules
        }
    })

    return deepClone
}
/**
 * Hook for input capturing and validation
 * 
 * Additional inputHandler functions should be added to handle more complex forms and edge case that are not handled
 * 
 * @param {object} defaultValues: Form Object to be initialised => { inputName: { value: '', validation: { touched: bool, required: bool, valid: bool, customChecks: fn } } }
 * See helper fromStore for object structure
 */
function useFormValidation(defaultValues) {
    const [form, setInput] = useState(cloneDefaultValues(defaultValues))

    // General input handler
    const handleInputChange = (event) => {

        // Our target attributes
        const target = event.target
        const value = target.checked ? target.checked : target.value
        const name = target.name

        // Make a copy of the form
        const updatedForm = { ...form }

        // Update the form
        const updatedFormElement = { ...updatedForm[name] }
        updatedFormElement.value = value
        updatedFormElement.validation.valid = checkValidation(updatedFormElement.value, updatedFormElement.validation)
        updatedFormElement.validation.touched = true
        updatedForm[name] = updatedFormElement

        setInput(updatedForm)
    }

    // Reset store to default valus
    const setDefaults = () => {
        setInput({ ...defaultValues })
    }

    // Format our form and return only the values
    const setFormSubmission = () => {
        const formData = {}

        // Loop through our form and get the values
        for (let formElementIdentifier in form) {
            formData[formElementIdentifier] = form[formElementIdentifier].value
        }

        return formData
    }

    const consumeValidation = (name) => {
        if (name) {

            // Make a copy of the form
            const updatedForm = { ...form }

            // Update the form
            const updatedFormElement = { ...updatedForm[name] }
            updatedFormElement.validation.valid = false
            updatedFormElement.validation.touched = false
            updatedForm[name] = updatedFormElement

            setInput(updatedForm)
        }

        return setInput(Object.keys(form).reduce((acc, key) => ({
            ...acc,
            [key]: {
                ...form[key],
                validation: {
                    ...form[key].validation,
                    touched: false,
                    valid: false,
                }
            }
        }), {}))
    }

    return [form, handleInputChange, setFormSubmission, setDefaults, consumeValidation]
}

/**
 * General validation function for form inputs
 * 
 * @param {string} value 
 * @param {object} validation: { required, customChecks: { p: fn, ... }} => customChecks should be a predicate fn else coercion will handle bool conversion
 */
export function checkValidation(value, validation) {
    const { required, customRules } = validation
    let isValid = true

    if (required) {
        isValid = value.trim() !== '' && isValid
    }

    if (customRules) {
        // loop all custom checks and make sure none are failing
        const passesAllTests = Object.values(customRules).every((check) => check(value))
        isValid = passesAllTests && isValid
    }

    return isValid
}

export default useFormValidation