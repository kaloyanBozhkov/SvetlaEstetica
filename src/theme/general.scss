@import './mixins.scss';
@font-face {
    font-family: 'SensationLight';
    src: local('SensationLight'), url('/fonts/Sansation_Light.ttf') format('truetype');
    font-style: normal;
    font-weight: 400;
}
@font-face {
    font-family: 'SensationRegular';
    src: local('SensationRegular'), url('/fonts/Sansation_Regular.ttf') format('truetype');
    font-style: bold;
    font-weight: 700;
}

html {
    //Defines what 1rem is: 1rem = 10px
    font-size: 62.5%; // 10px / 16px = 62.5% *
    line-height: 1.15; /* 1 */
    -webkit-text-size-adjust: 100%; /* 2 */
    // Example of how we could handle resizing the entire app just by handling the root font size from here :)

    // @include breakpoint(tablet-landscape) {
    //     font-size: 56.25%; //1rem = 9px;
    // }

    // @include breakpoint(tablet-portrait) {
    //     font-size: 50%; //1rem = 8px; 8/16 = 50%
    // }

    // @include breakpoint(mobile) {
    //     font-size: 37.5%; //1rem = 6px; 6/16 = 37.5%
    // }

    // @include breakpoint(big-desktop) {
    //     font-size: 75%; //1 rem = 12px;
    // }
}

html,
body,

// @TODO Koko figure out why #__next is not being targeted, and we gotta do style jsx global on _document instead
#__next {
    height: 100%;
    width: 100%;
    overflow: auto;
}

select {
    option {
        color: #4B4B4B
    }
}