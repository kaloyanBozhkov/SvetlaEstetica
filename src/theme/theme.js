import { theme } from '@chakra-ui/core'

// Chakra use mobile-first design so it uses min-widths breakpoints
/**
 * 20em: 320px small mobile
 * 30em: 480px mobile
 * 40em: 640px large phone / tablet
 * 56em: 900px large tablet
 * 80em: 1280px Standard deskop/laptop
 * 112.5em: 1800px large desktop
 */
const breakpoints = ['20em', '30em', '40em', '56em', '80em', '112.5em']

/**
 * REM TO PX CONVERSION
 * 
 * DESKTOP: 1rem = 10Px
 */

// aliases
breakpoints.xs = breakpoints[0]
breakpoints.sm = breakpoints[1]
breakpoints.md = breakpoints[2]
breakpoints.mdX = breakpoints[3]
breakpoints.base = breakpoints[4]
breakpoints.lg = breakpoints[5]

// Import Chakra base theme and overide and set our variables
export const localTheme = {
    ...theme,
    breakpoints: breakpoints,
    colors: {
        ...theme.colors,
        main: {
            900: '',
            800: '#EA6587',
            700: '#BC808C',
            600: '',
            500: '',
            400: '',
            300: '#FFF3F7',
            200: '',
            100: '#F7F5DD'
        },
        background: {
            800: '#4B4B4B',
            200: '#C1C1C1',
            100: '#e4e4e4'
        },
        neutral: {
            900: '#4B4B4B',
            800: '',
            700: '',
            600: '',
            500: '',
            400: ''
        },
        green: {
            900: '#0EE56F',
            800: '#57B16A',
            700: '#8EC434'
        },
        red: {
            900: '#D0021B',
            350: '#D44469',
            300: '#EA6587'
        },
        text: {
            900: '#4B4B4B',
            400: '#BBBBBB',
            300: '#e2e2e2'
        },
        white: '#FFFFFF'

    },
    fonts: {
        body: 'SensationLight, SensationRegular, Helvetica Neue, -apple-system, BlinkMacSystemFont, Helvetica, Arial, sans-serif',
        heading: 'SensationRegular, SensationLight, Helvetica Neue',
        mono: 'SensationLight, SensationRegular, Menlo, monospace',
    },
    fontSizes: {
        ...theme.fontSizes,
        xxs: '10px',
        xs: '12px',
        sm: '14px',
        md: '16px',
        lg: '18px',
        xl: '30px',
        '2xl': '38px',

    },
    fontWeights: {
        normal: 300,
        medium: 400,
        bold: 700
    },
    lineHeights: {
        tighter: '18px',
        tight: '22px',
        normal: '0',
        wide: '24px',
        wider: '32px',
        widest: '50px'
    },
    sizes: {
        ...theme.sizes,
        0: '0',
        1: '1rem',
        2: '2rem',
        3: '3rem',
        4: '4rem',
        5: '5rem',
        6: '6rem',
        '6xl': '6.8rem',
        12: '12rem',
        24: '10rem',
        13: '13rem',
        60: '44rem',
        'sideMenu': '30rem'
    },
    space: {
        ...theme.space,
        0: '0px',
        px: '1px',
        1: '0.5rem',
        2: '1rem',
        3: '1.2rem',
        4: '1.6rem',
        5: '2rem',
        6: '2.4rem',
        8: '3.2rem',
        10: '4rem',
        12: '4.8rem',
        16: '6rem',
        20: '8rem',
        24: '10rem',
        32: '12.8rem',
        40: '16rem',
        48: '19.2rem',
        50: '21.6rem',
        56: '22.4rem',
        64: '25.6rem'
    },
    zIndices: {
        ...theme.zIndices,
        hide: -1,
        auto: 'auto',
        base: 0,
        docked: 10,
        dropdown: 1000,
        sticky: 1100,
        banner: 1200,
        overlay: 1300,
        modal: 1400,
        popover: 1500,
        skipLink: 1600,
        toast: 1700,
        tooltip: 1800,
    },
    shadows: {

    }
}

export const pagePaddings = {
    paddingLeft: {
        xs: '2',
        sm: '4',
        md: '8',
        base: '50'
    },
    paddingRight: {
        xs: '2',
        sm: '4',
        md: '8',
        base: '50'
    },
}