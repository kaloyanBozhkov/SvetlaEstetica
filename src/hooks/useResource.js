import useDataFetch from './useDataFetch'

// allows to handle data/loading/erorr with ErrorBoundary and Suspense instead of variables returned
const useResource = (promiseFn, checker) => {
    const [response,, error, performDataFetch] = useDataFetch(promiseFn, checker)

    return {
        read() {
            if (error) throw error
            if (response) return response
            throw promiseFn
        },
        refetch(...args) {
            performDataFetch(...args)
        },
    }
}
 
export default useResource