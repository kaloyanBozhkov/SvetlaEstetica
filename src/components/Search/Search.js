import { Button, Flex, Grid, Input } from '@chakra-ui/core'
import React, { useEffect } from 'react'

// import hooks
import useFormValidation from '~/hooks/useFormValidation'
import useEscapeKeyPressEvent from '~/hooks/useEscapeKeyPressEvent'

// import components
import Icon from 'UI/Icon/Icon'
import { SpinnerCircular } from 'spinners-react'

const styles = {
    contianer: {
        bg: 'white',
        paddingTop: '22'
    },
    input: {
        height: '8rem',
        border: 'none',
        textTransform: 'capitalize'
    },
    searchWrapper: {
        borderBottom: '1px solid',
        borderBottomColor: 'neutral.100',
        h: 'min-content',
        w: '100%'
    }
}

let timer = null

const initialInputs = {
    search: {
        value: '',
        validation: {
            required: true,
            touched: false,
            valid: false,
            customRules: {
                isMinLength: (v) => v.toString().length > 2
            }
        }
    }
}

const Search = ({ searchActive = false, isLoading = false, onClearSearch = (f) => f, onSearch = (f) => f, onCloseSearch = (f) => f }) => {
    // hook controlling escape on window to close search
    useEscapeKeyPressEvent(onCloseSearch)

    const [inputs, handleInputChange, onSubmit, onSetDefaults, consumeValidation] = useFormValidation(initialInputs)

    // handle search by input
    useEffect(() => {
        // if input is valid (passes its customRules)
        if (inputs.search.validation.valid) {

            const value = onSubmit().search

            // if timer is already counting, restart it!
            if (timer) window.clearTimeout(timer)

            timer = window.setTimeout(() => {
                onSearch(value)

                // after running this block, consume validation so to allow it to run only after on input makes it pass validation again
                consumeValidation()
            }, 500)

            // on unmount be sure to clear
            return () => window.clearTimeout(timer)
        }
    }, [inputs.search, onSearch, onSubmit, consumeValidation])

    const clearSearchClick = () => {
        if (searchActive) onClearSearch()
        onSetDefaults(initialInputs)
    }

    return (
        <Flex {...styles.contianer}>
            <Flex {...styles.searchWrapper}>
                <Input data-test-id="input" placeholder="search by typing.. (write 'test' to fall back to stubs)" name="search" value={inputs.search.value} onChange={handleInputChange} {...styles.input} />
                {(searchActive || !isLoading) && (
                    <Flex alignItems="center" padding="2">
                        <Button data-test-id="clearSearchBtn" bg="none" onClick={clearSearchClick}>clear search</Button>
                    </Flex>
                )}
                <Flex padding="4" alignItems="center">
                    {isLoading ? (
                        (<SpinnerCircular data-test-id="spinner" size="18" color="black" thickness="135" />)
                    ) : (<Button data-test-id="closeSearchBtn" bg="none" onClick={onCloseSearch}><Icon icon="times" /></Button>)}
                </Flex>
            </Flex>
        </Flex>
    )
}

export default Search