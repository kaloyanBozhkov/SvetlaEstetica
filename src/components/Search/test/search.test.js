import { mount, shallow } from 'enzyme'
import React from 'react'
import Search from '../Search'
import * as useEscapeKeyPressEvent from '~/hooks/useEscapeKeyPressEvent'
import * as useFormValidation from '~/hooks/useFormValidation'
import { ThemeProvider } from 'emotion-theming'
import { localTheme } from 'theme/theme'

describe('tesitng search component', () => {

    beforeEach(() => {
        jest.useFakeTimers()
    })

    const mockProps = {
        searchActive: false,
        isLoading: false,
        onClearSearch: jest.fn((f) => f),
        onSearch: jest.fn((f) => f),
        onCloseSearch: jest.fn((f) => f)
    }

    it('should render component w default params', () => {
        const wrapper = shallow(<Search />)
        expect(wrapper).toMatchSnapshot()
    })

    it('should trigger close btn', () => {
        const wrapper = shallow(<Search {...mockProps} />)
        const closeBtn = wrapper.find('[data-test-id="closeSearchBtn"]')
        expect(mockProps.onCloseSearch).toHaveBeenCalledTimes(0)
        closeBtn.simulate('click')
        expect(mockProps.onCloseSearch).toHaveBeenCalledTimes(1)
    })

    it('should handle setting input, trigger clear btn & reset input val', () => {
        const wrapper = shallow(<Search {...mockProps} />)
        let clearBtn = wrapper.find('[data-test-id="clearSearchBtn"]')
        let input = wrapper.find('[data-test-id="input"]')

        // initially notcalled
        expect(mockProps.onClearSearch).toHaveBeenCalledTimes(0)

        // initially empty
        expect(input.prop('value')).toEqual('')

        // set some vlaue
        input.prop('onChange')({ target: { value: 'new val', name: 'search' } })

        // set active search prop + update wrapper
        wrapper.setProps({ searchActive: true })

        input = wrapper.find('[data-test-id="input"]')

        // updatedvalue
        expect(input.prop('value')).toEqual('new val')
        clearBtn = wrapper.find('[data-test-id="clearSearchBtn"]')

        clearBtn.prop('onClick')()

        wrapper.update()

        input = wrapper.find('[data-test-id="input"]')

        // all has been reset!
        expect(mockProps.onClearSearch).toHaveBeenCalledTimes(1)
        expect(input.prop('value')).toEqual('')
    })

    it('should run hook fn on mount', () => {
        const spy = jest.spyOn(useEscapeKeyPressEvent, 'default')
        const mockFn = jest.fn((f) => f)
        spy.mockImplementation(mockFn)
        // has not  been called yet
        expect(mockFn).toHaveBeenCalledTimes(0)
        const wrapper = shallow(<Search {...mockProps} />)
        // has been called after initialization
        expect(mockFn).toHaveBeenCalledTimes(1)
        expect(mockFn.mock.results[0].value).toEqual(mockProps.onCloseSearch)
    })

    it('should show loading spinner if isLoading', () => {
        let wrapper = shallow(<Search {...mockProps} />)
        let spinner = wrapper.find('[data-test-id="spinner"]')
        expect(spinner.length).toEqual(0)
        wrapper = shallow(<Search {...mockProps} isLoading />)
        spinner = wrapper.find('[data-test-id="spinner"]')
        expect(spinner.length).toEqual(1)
    })

    it('should run search fn on input of >3 chars after 500ms', () => {
        const spy = jest.spyOn(useFormValidation, 'default')
        const overridenInputValue = {
            search: {
                value: 'some value to search',
                validation: {
                    required: true,
                    touched: false,
                    valid: true,
                    customRules: {
                        isMinLength: (v) => v.length > 2
                    }
                }
            }
        }

        const mocks = {
            handleChange: jest.fn(f => f),
            submit: jest.fn((f) => ({ search: overridenInputValue.search.value })),
            consume: jest.fn((f) => f),
            setDefaults: jest.fn((f) => f)
        }

        const mockUseFormValidation = jest.fn((initialInputVal) => [
            overridenInputValue,
            mocks.handleChange,
            mocks.submit,
            mocks.setDefaults,
            mocks.consume
        ])

        spy.mockImplementation(mockUseFormValidation)

        const wrapper = mount(
            <ThemeProvider theme={localTheme}>
                <Search {...mockProps} />
            </ThemeProvider>
        )

        // initialized hook
        expect(mockUseFormValidation).toHaveBeenCalledTimes(1)

        // input set to overriden value on start already
        expect(wrapper.find('Input').prop('value')).toEqual(overridenInputValue.search.value)

        // useEffect runs, input has value and is valid so the following run too
        expect(mocks.submit).toHaveBeenCalledTimes(1)
        expect(mocks.setDefaults).toHaveBeenCalledTimes(0)
        
        // consume validation only after search
        expect(mocks.consume).toHaveBeenCalledTimes(0)
        
        // initially onSearch not called
        expect(mockProps.onSearch).toHaveBeenCalledTimes(0)

        // run the timer callback
        jest.runOnlyPendingTimers()

        // search has ran!
        expect(mockProps.onSearch).toHaveBeenCalledTimes(1)

        // consume validation only after search
        expect(mocks.consume).toHaveBeenCalledTimes(1)
        
    })
})