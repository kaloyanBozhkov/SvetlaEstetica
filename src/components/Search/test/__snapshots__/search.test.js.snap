// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tesitng search component should render component w default params 1`] = `
<Flex
  bg="white"
  paddingTop="22"
>
  <Flex
    borderBottom="1px solid"
    borderBottomColor="neutral.100"
    h="min-content"
    w="100%"
  >
    <Input
      as="input"
      border="none"
      data-test-id="input"
      errorBorderColor="red.500"
      focusBorderColor="blue.500"
      height="8rem"
      isFullWidth={true}
      name="search"
      onChange={[Function]}
      placeholder="search by typing.. (write 'test' to fall back to stubs)"
      size="md"
      textTransform="capitalize"
      value=""
      variant="outline"
    />
    <Flex
      alignItems="center"
      padding="2"
    >
      <Button
        bg="none"
        data-test-id="clearSearchBtn"
        onClick={[Function]}
      >
        clear search
      </Button>
    </Flex>
    <Flex
      alignItems="center"
      padding="4"
    >
      <Button
        bg="none"
        data-test-id="closeSearchBtn"
        onClick={[Function]}
      >
        <Icon
          icon="times"
        />
      </Button>
    </Flex>
  </Flex>
</Flex>
`;
